pkg/kptfile/kptfileutil/util.go

```diff
@@ -261,6 +271,85 @@ func merge(localKf, updatedKf, originalKf *kptfilev1.KptFile) error {
 	return nil
 }
 
+// shouldAddSyntheticMergeName returns true iff all the functions in all the sources
+// doesn't have name field set, it means the user is unaware of name field, and we
+// use image as mergeKey instead of name in such cases
+// it also checks if there are no duplicate image name declarations
+func shouldAddSyntheticMergeName(kfs ...*kptfilev1.KptFile) bool {
+	for _, kf := range kfs {
+		if kf == nil || kf.Pipeline == nil {
+			continue
+		}
+		if !shouldAddSyntheticMergeNameUtil(kf.Pipeline.Mutators) || !shouldAddSyntheticMergeNameUtil(kf.Pipeline.Validators) {
+			return false
+		}
+	}
+	return true
+}
+
+// shouldAddSyntheticMergeNameUtil is util method for shouldAddSyntheticMergeName
+func shouldAddSyntheticMergeNameUtil(fns []kptfilev1.Function) bool {
+	imageNamesSet := sets.String{}
+	for _, fn := range fns {
+		if fn.Name != "" {
+			return false
+		}
+		imageName := strings.Split(fn.Image, ":")[0]
```

Comment:
Since support for `exec` is merged now, we need to handle that as well. For `exec`, we might have to parse the execString to extract the executable binary and that will become the fnKey I guess.